File: AdventOfCode2018/6/csharp/Program.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace csharp {

    class Point {
        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
        public int distance(Point p) {
            return Math.Abs(x - p.x) + Math.Abs(y - p.y);
        }
        public int x, y;
    }

    class Program {

        static Point[] FindBoundingBox(List<Point> points) {
            int? minX = null, minY = null, maxX = null, maxY = null;
            foreach (var p in points) {
                if (p.x > maxX || maxX == null) maxX = p.x;
                if (p.x < minX || minX == null) minX = p.x;
                if (p.y > maxY || maxY == null) maxY = p.y;
                if (p.y < minY || minY == null) minY = p.y;
            }
            Point[] boundingBox = {
                new Point(minX ?? -1, minY ?? -1),
                new Point(maxX ?? -1, maxY ?? -1)};
            return boundingBox;
        }

        static string Part1(List<Point> input) {
            var BB = FindBoundingBox(input);
            var AreaForPoint = new Dictionary<Point, int>();

            for (var x = BB[0].x; x < BB[1].x; x++) {
                for (var y = BB[0].y; y < BB[1].y; y++) {
                    int? minDistance = null;
                    Point closetPoint = null;
                    Point currentCoord = new Point(x, y);
                    foreach (var p in input) {
                        // terminate for same distance
                        if (minDistance != null && p.distance(currentCoord) == minDistance) {
                            closetPoint = null;
                            break;
                        }
                        if (minDistance == null || p.distance(currentCoord) < minDistance) {
                            minDistance = p.distance(currentCoord);
                            closetPoint = p;
                        }
                    }
                    if (closetPoint == null)
                        continue;
                    if (!AreaForPoint.ContainsKey(closetPoint))
                        AreaForPoint[closetPoint] = 0;
                    AreaForPoint[closetPoint]++;
                }
            }
            return AreaForPoint.Values.Max().ToString();
        }

        static string part2(List<Point> input) {
            var AreaUnderThreshold = 0;
            var BB = FindBoundingBox(input);
            for (var x = BB[0].x; x < BB[1].x; x++) {
                for (var y = BB[0].y; y < BB[1].y; y++) {
                    var currentPoint = new Point(x, y);
                    var totalDistance = input.Sum(p => p.distance(currentPoint));
                    if (totalDistance < 10000)
                        AreaUnderThreshold++;
                }
            }
            return AreaUnderThreshold.ToString();
        }

        static void Main(string[] args) {
            List<Point> input = new List<Point>();
            foreach (var line in System.IO.File.ReadLines(args[0])) {
                var m =  Regex.Match(line, @"(\d+), (\d+)");
                var t = new Point(Int32.Parse(m.Groups[1].Value), Int32.Parse(m.Groups[2].Value));
                input.Add(t);
            }

            // part 1 and 2
            Console.WriteLine(Part1(input));
            Console.WriteLine(part2(input));
        }
    }
}

File: AdventOfCode2018/6/fsharp/Program.fs
open System
open System.Text.RegularExpressions

type Point = {x:int; y:int} with
    static member New x y = {x = x; y = y}
    static member Origin = Point.New 0 0
    static member Parse line =
        let m = Regex.Match(line, "(\d+), (\d+)")
        {x = int(m.Groups.[1].Value); y=int(m.Groups.[2].Value)}
    member this.distanceTo p =
        abs(this.x - p.x) + abs(this.y - p.y)

let nearestPoint (p:Point) points =
    let mutable minDistance = None
    let mutable sameDistance = false
    let mutable minId = None
    points |> Seq.indexed |> Seq.iter (fun (id, px) ->
        let distance = p.distanceTo px
        match minDistance with
        | None -> minDistance <- Some distance; minId <- Some id
        | Some d when d > distance -> minDistance <- Some distance; minId <- Some id
        | Some d when d = distance -> sameDistance <- true
        | _ -> ()
    ) 
    if sameDistance then None else minId
    
let boundingBox points = 
    points
    |> Seq.fold (fun (left, right) p ->
        let lx = if left.x < p.x then left.x else p.x
        let ly = if left.y < p.y then left.y else p.y
        let rx = if right.x > p.x then right.x else p.x
        let ry = if right.y > p.y then right.y else p.y
        (Point.New lx ly, Point.New rx ry)
    ) (Point.New 1000 1000, Point.Origin)

let idsOnBoundaryBox points (min, max) =
    let vertPoints x yRange = Seq.map (fun y -> {x = x; y = y}) yRange
    let horzPoints xRange y = Seq.map (fun x -> {x = x; y = y}) xRange
    let top = horzPoints [min.x..max.x] min.y
    let bottom = horzPoints [min.x..max.x] max.y
    let left = vertPoints min.x [min.y..max.y]
    let right = vertPoints max.x [min.y..max.y]

    Seq.concat [top; bottom; left; right]
    |> Seq.map (fun p -> nearestPoint p points)
    |> Seq.distinct
    |> Seq.filter (fun p -> p.IsSome)
    |> Seq.map (fun (Some p) -> p)

let part1 points =
    let (topLeft, bottomRight) = boundingBox points
    let sizePerId = Array.zeroCreate (points |> Seq.length)
    let idsOfInfiniteAreas = idsOnBoundaryBox points (topLeft, bottomRight)
    printf "%A" topLeft
    printf "%A" bottomRight
    for x = topLeft.x to bottomRight.x do
        for y = topLeft.y to bottomRight.y do
            let currentLoc = Point.New x y
            match nearestPoint currentLoc points with
            | Some id -> sizePerId.[id] <- sizePerId.[id] + 1
            | _ -> ()

    Array.indexed sizePerId
    |> Array.filter (fun (id, _) -> not(Seq.contains id idsOfInfiniteAreas))
    |> Array.maxBy snd |> snd

let part2 points =
    let (min, max) = boundingBox points
    let mutable counts = 0
    for x = min.x to max.x do
        for y = min.y to max.y do
            let currentPos = Point.New x y
            let totalDistance = points |> Seq.map currentPos.distanceTo |> Seq.sum
            if totalDistance < 10000 then counts <- counts + 1 else ()
    counts

[<EntryPoint>]
let main argv =
    let points = System.IO.File.ReadLines argv.[0] |> Seq.map Point.Parse
    part1 points |> printf "%A\n"
    part2 points |> printf "%A\n"
    0
